File: subprojs/tradeapi/app/scala/MarketChannel.scala
package org.kangmo.tradeapi

import org.kangmo.http._
import org.kangmo.helper._

import java.math.BigDecimal

case class Ticker (	
	timestamp: java.sql.Timestamp,
	last: BigDecimal
)

case class FullTicker (	
	timestamp: java.sql.Timestamp,
	last: BigDecimal,
	bid: Option[Long],
	ask: Option[Long],
	low: Option[BigDecimal],
	high: Option[BigDecimal],
	volume: Option[BigDecimal]
)

case class OrderBook (
	timestamp: java.sql.Timestamp,
	// The bid price represents the maximum price that a buyer or buyers are willing to pay
	bids : Seq[Seq[BigDecimal]],
	// The ask price represents the minimum price that a seller or sellers are willing to receive
	asks : Seq[Seq[BigDecimal]]
)

case class Transaction (
	timestamp: Option[java.sql.Timestamp],
  date: Option[Long],
	tid: Long,
	price: BigDecimal,
	amount: BigDecimal
)

case class TransactionId(id: Long)

class MarketChannel extends AbstractChannel {
	def ticker() = getPublicFuture[Ticker]("ticker")

	def fullTicker() = getPublicFuture[FullTicker]("ticker/detailed")

	def orderbook() = getPublicFuture[OrderBook]("orderbook?group=true")

	def transactions(sinceTransactionId : TransactionId) = {
		getPublicFuture[Seq[Transaction]](s"transactions?since=${sinceTransactionId.id}")
	}
}

File: src/main/kotlin/org/kangmo/tradeapi/MarketChannel.kt
package org.kangmo.tradeapi

import java.math.BigDecimal

data class Ticker (
	val timestamp: java.sql.Timestamp,
	val last: BigDecimal
)

data class FullTicker (
	val timestamp: java.sql.Timestamp,
	val last: BigDecimal,
	val bid: Long?,
	val ask: Long?,
	val low: BigDecimal?,
	val high: BigDecimal?,
	val volume: BigDecimal?
)

data class OrderBook (
	val timestamp: java.sql.Timestamp,
	// The bid price represents the maximum price that a buyer or buyers are willing to pay
	val bids : List<List<BigDecimal>>,
	// The ask price represents the minimum price that a seller or sellers are willing to receive
	val asks : List<List<BigDecimal>>
)

data class Transaction (
	val timestamp: java.sql.Timestamp?,
	val tid: Long,
	val price: BigDecimal,
	val amount: BigDecimal
)

data class TransactionId(val id: Long)


class MarketChannel: AbstractChannel() {
	suspend fun ticker(currencyPair : CurrencyPair = CurrencyPair.BtcKrw) = getPublicFuture<Ticker>("ticker?currency_pair=${currencyPair.pair}", Ticker::class.java)

	suspend fun fullTicker(currencyPair : CurrencyPair = CurrencyPair.BtcKrw) = getPublicFuture<FullTicker>("ticker/detailed?currency_pair=${currencyPair.pair}", FullTicker::class.java)

	suspend fun orderbook(currencyPair : CurrencyPair = CurrencyPair.BtcKrw) = getPublicFuture<OrderBook>("orderbook?currency_pair=${currencyPair.pair}", OrderBook::class.java)

	suspend fun transactions(currencyPair : CurrencyPair = CurrencyPair.BtcKrw, interval : TimeInterval) : List<Transaction> =
		getPublicFuture<List<Transaction>>("transactions?currency_pair=${currencyPair.pair}&time=${interval.unit}", List::class.java as Class<List<Transaction>>)
}